Adding initial blog post content (#703)
Quansight/Quansight-website: apps/consulting/posts/openai-pseudocode.md

---
title: 'Collaborative code generation with LLMs'
published: March 27, 2023
author: chris-ostrouchov
description: 'As an experiment we developed a python package pseudocode which much like its name implies allows developers to annotate their code with pseudocode and produce testable code generated by LLMs without the user writing any code.'

> category: ["ML", "LLM", "python", "openai"]

Reviewer: Likely will want to include at least one of the [already-existing categories][categories.json] so that the "see these other related articles" component at the bottom of the blog page will show some posts. `Artificial Intelligence` seems the most relevant to me.

Any categories you wish to add here that don't exist yet in [`categories.json`][categories.json] will need to be added there, following the object format defined there.

[categories.json]: https://github.com/Quansight/Quansight-website/blob/develop/apps/consulting/posts/categories.json

Reviewer 2: @costrouc add AI as suggested and modify `categories.json`

Author: Found out there is already an `artificial-intelligence` category. Used that.

featuredImage:
  src: '/posts/openai-pseudocode/pseudocode-screenshot.png'
  alt: 'A screenshot of the pseudocode program running'
hero:
  imageSrc: /posts/hero-paris.webp
  imageAlt: 'Data visualization of Paris city'
---

<base target="_blank" />

## Overview

Large Language Models (LLMs) have revolutionized the field of natural
language processing (NLP) and artificial intelligence (AI), opening up
a world of possibilities across various domains. These models have
been trained on vast amounts of data and are capable of generating
human-like responses to a wide range of queries. The developer
community has seen especially good results due to the large amount of
source code available online for training.

Github's Co-pilot is a tool which will autocomplete code for a

> developer but lacks the ability to interact with the completation,

Reviewer 3: ```suggestion
developer but lacks the ability to interact with the completion,
```

> developer but lacks the ability to interact with the completation,

Reviewer 4: ```suggestion
developer but lacks the ability to interact with the generated code,
```

provide feedback, and enforce a test suite on the generated code. We
think this iterative loop is a core feature of LLM's which need to be
leveraged.

> As an experiment we developed a python package

Reviewer 4: ```suggestion
As an experiment we developed a Python package
```

[pseudocode](https://github.com/Quansight/pseudocode) which allows

> developers to annotate their code and produce testable code generated

Reviewer 4: ```suggestion
developers to annotate their code and produce tested code generated
```

by LLMs without the user writing any code. We hope that
[pseudocode](https://github.com/Quansight/pseudocode) may be a "higher

> level language" for writing python code. Below you will find an

Reviewer 4: ```suggestion
level language" for writing Python code. Below you will find an
```

example. We emphasize interfaces via type annotations and function

> docstrings and provide easy ways to have automated tests. 

Reviewer 4: ```suggestion
docstrings and provide easy ways to include automated tests. 
```

## Example

Let's use a concrete example of a slightly non-trivial task of getting

> Github issues.

Reviewer 4: ```suggestion
GitHub issues.
```

```python
import typing
import datetime

from pseudocode import pseudo_function

@pseudo_function(review=True)
def get_issues(respository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]:
    """A function to fetch all issues created by

    A function that does the following:
     - assume you have a github token environment variable GITHUB_TOKEN
     - assume that repositiory is of the form organization/repo
     - use the requests library
     - fetch all github issues from repository in last 10 days
     - only show issue numbers which are odd
     - return a tuple with issue titles, number, and date created

    Examples
    --------
    >>> all([_[1] % 2 == 1 for _ in get_issues("conda/conda")])
    True
    """
    pass

print('github issues', get_issues("conda/conda"))
```

We feel this is a good example because it is hard to remember how to

> use the github API exactly and inevitably requires some back and forth

Reviewer 4: ```suggestion
use the GitHub API exactly, and inevitably requires some back and forth
```

using the [PyGithub](https://github.com/PyGithub/PyGithub) python

> library, requests, and viewing API documentation. LLMs show promise

Reviewer 4: ```suggestion
library, Requests, and viewing API documentation. LLMs show promise
```

for these applications since there is a large amount of surrounding
documentation but no one example to do exactly what you
need. `pseudocode` enforces the practice of creating an __interface__
specification which:
 - declares the function signature `def get_issues(respository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]`
 - uses the function's `docstring` to specify what the function should be doing

>  - automated tests run by `pseudocode` to provide feedback to the LLM of the quality and correctness of the code generated

Reviewer 4: ```suggestion
 - automates tests run by `pseudocode` to provide feedback to the LLM of the quality and correctness of the code generated
```

> When you run the following code you will see the following.

Reviewer 4: ```suggestion
When you run the example above, you will see the following.
```

│     response = requests.get(url, headers=headers, params=params)                                 │
│     issues = response.json()                                                                     │
│     return [(i['title'], i['number'], i['created_at']) for i in issues if i['number'] % 2 != 0]  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
What code feedback would you like to provide? (leave empty for approval):
```

Already after some quick initial feedback we can see that the LLM
takes into account the users needs. After user provided feedback the
automated tests are run. If those tests fail or run into exceptions
they will be automatically passed back to the LLM for corrections.

```
pass result = str(all([_[1] % 2 == 1 for _ in run("conda/conda")]))
github issues [
  ('Use appropriate `defaults` value for Windows is use when subdir=win-*', 12555, '2023-03-30T17:44:55Z'), 
  ...
  ('Document steps to debug tests with vscode', 12459, '2023-03-07T18:20:08Z')
]
```

> 

Reviewer 4: I expect to see the final code at this point.

## Conclusion

There are several key ideas here we want to highlight about why we
think the approaches taken in `pseudocode` are unique.

> 

Reviewer 4: The first point "Autogenerated code needs guidance" is clear and should be at the start of the bullet.  The second two bullets need similar style in which the bullet starts with a clear point.  "LLMs need declared interfaces"?  "Code declarations and generated code should be separated"?

 * Feedback is generated from the user and the rest is guided by
   automated testing and exception reporting back to the
   LLM. __Autogenerated code needs guidance__.

 * Declaring interfaces for LLMs to operate within. Interface design
   is naturally a high level task which is the key to composable code.

 * Separation of code declarations and generated code. Similar to
   `.py` and `.pyc` files.

> We mentioned earlier that this was an experiment of the interaction

Reviewer 4: ```suggestion
I mentioned earlier that this was an experiment of the interaction
```

> between automated testing, user feedback, and LLMs. We want to explore

Reviewer 4: Change to "I" throughout if there is only one author.

```suggestion
between automated testing, user feedback, and LLMs. I want to explore
```

> this more and already have additional ideas we would like to

Reviewer 4: "...additional ideas, like applying linting to generated LLM code via black, ..."

explore. We would like to apply formatting and linting to generated
LLM code via [black](https://black.readthedocs.io/en/stable/),
[ruff](https://github.com/charliermarsh/ruff), and
[isort](https://pycqa.github.io/isort/). Caching of LLM code generated
to avoid repetitive API calls and a database of reusable code. Similar
to `pseudocode.pseudo_function(...)` it would be nice to have an
equivalent for arbitrary files. This feels somewhat like
[cookiecutter](https://cookiecutter.readthedocs.io/en/stable/) on

> steriods

Reviewer 4: ```suggestion
steroids.
```

---

> title: 'Build your own code copilot with LLMs'

Reviewer: Should be title case to match Consulting site branding guide.

```suggestion
title: 'Build Your Own Code Copilot With LLMs'
```

> published: March 27, 2023

Reviewer: This will probably go live Wednesday, if we keep the current planned schedule.

```suggestion
published: April 5, 2023
```

author: chris-ostrouchov

> description: 'As an experiment we wanted to add a feedback loop into LLM generated code via type annotations, docstrings, automated tests, and user feedback. We developed a simple package pseudoscode which uses the OpenAI API for generating code.'

Reviewer: This should be set off to show it's a name. It'll be rendered plaintext, so backticks won't get code formatting. I suggest double quotes.

```suggestion
description: 'As an experiment we wanted to add a feedback loop into LLM generated code via type annotations, docstrings, automated tests, and user feedback. We developed a simple package, "pseudoscode," which uses the OpenAI API for generating code.'
```

Reviewer 4: I thought we used italics in place of backticks during the migration.

> category: ["artificial-intelligence"]

Reviewer: This needs to match the `name` property of the category object, and should not have enclosing double quotes.

```suggestion
category: [Artificial Intelligence]
```

featuredImage:
  src: '/posts/openai-pseudocode/pseudocode-screenshot.png'
  alt: 'A screenshot of the pseudocode program running'
hero:
  imageSrc: /posts/hero-paris.webp
  imageAlt: 'Data visualization of Paris city'
---

<base target="_blank" />

## Overview

Large Language Models (LLMs) have been around for several years but
recent advances have revolutionized the field of natural language
processing (NLP) and artificial intelligence (AI), opening up a world
of possibilities across various domains. [OpenAI's
ChatGPT](https://chat.openai.com/) has taken the world by storm and
shown remarkable ability to generate human-like responses to a wide
range of queries.

These models require training on vast amounts of data and in some
sense can be thought of as a way to summarize and expose the
information in the training dataset. It is important to remember
however that the information returned in response to a query is a
statistical answer and these models have no concept of

> 'correctness'. This leads to the so called hallucination effect that

Reviewer: ```suggestion
'correctness.' This leads to the so called hallucination effect that
```

Reviewer 4: We have a suggested new intro in the comments, which would replace this sentence.
https://github.com/Quansight/Quansight-website/pull/703#issuecomment-1494634773

> is now being talked about, i.e. when a confident but factually wrong

Reviewer: ```suggestion
is now being talked about, i.e., when a confident but factually wrong
```

response is given to a set of queries.

Between discussion sites like StackOverflow and code repositories like
GitHub and GitLab, there is a tremendous amount of source code
available online for potential training. Efforts like [Github
Copilot](https://github.com/features/copilot), show the power of AI
assisted coding to boost productivity. Our testing has shown however

> that while CoPilot can greatly assist a developer, it is also prone to

Reviewer: Google only capitalizes the 'C'

```suggestion
that while Copilot can greatly assist a developer, it is also prone to
```

hallucinations inventing new methods that don't exist or sometimes

> using outdated api's that no longer exist in recent versions of the

Reviewer 5: APIs

code.

We decided to see if we could add a feedback loop into the process of
code generation to reduce model hallucinations and iterate towards a

> working solution faster. We also wanted to experiment with adding

Reviewer: ```suggestion
working solution more quickly. We also wanted to experiment with adding
```

conversational feedback to the results rather than the auto-complete

> mechanism that CoPilot currently implements (Note: CoPilot just

Reviewer: ```suggestion
mechanism that Copilot currently implements (Note: Copilot just
```

announced a preview release of a [conversational chat based

> mode](https://github.blog/2023-03-22-github-copilot-x-the-ai-powered-developer-experience/).

Reviewer: Need an extra close paren to close the parenthetical.

```suggestion
mode](https://github.blog/2023-03-22-github-copilot-x-the-ai-powered-developer-experience/)).
```

> As an experiment we developed a python package

Reviewer 5: Python

> As an experiment we developed a Python package

Reviewer: ```suggestion
As an experiment we developed a Python package,
```

[pseudocode](https://github.com/Quansight/pseudocode) which allows
developers to annotate their code and produce testable code generated
by LLMs without the user writing any code. We hope that
[pseudocode](https://github.com/Quansight/pseudocode) may be a "higher

> level language" for writing python code. Below you will find an

Reviewer 5: Python

example. We emphasize interfaces via type annotations and function
docstrings and provide easy ways to have automated tests.

1. We define the code to be generated as a function signature with
   type annotations
2. We define tests that must pass

> 3. We submit these to OpenAI `gpt-3.5-turbo` API with some

Reviewer: ```suggestion
3. We submit these to the OpenAI `gpt-3.5-turbo` API with some
```

   resubmit the errors back to OpenAI `gpt-3.5-turbo`

Below we walk through a demonstration of this process. We have named
the codebase [pseudocode](https://github.com/Quansight/pseudocode).

## Example

Let's use a concrete example of a slightly non-trivial task of getting
Github issues.

```python
import typing
import datetime

from pseudocode import pseudo_function

@pseudo_function(review=True)
def get_issues(repository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]:
    """A function to fetch all issues created by

    A function that does the following:
     - assume you have a github token environment variable GITHUB_TOKEN
     - assume that repository is of the form organization/repo
     - use the requests library
     - fetch all github issues from repository in last 10 days
     - only show issue numbers which are odd
     - return a tuple with issue titles, number, and date created

    Examples
    --------
    >>> all([_[1] % 2 == 1 for _ in get_issues("conda/conda")])
    True
    """
    pass

print('github issues', get_issues("conda/conda"))
```

We feel this is a good example because it is hard to remember how to

> use the GitHub API exactly and inevitably requires some back and forth

Reviewer: ```suggestion
use the GitHub API exactly, and successful use inevitably requires some back and forth
```

using the [PyGithub](https://github.com/PyGithub/PyGithub) Python

> library, API requests, and viewing API documentation. LLMs show promise

Reviewer: This sentence should be de-split across a paragraph break.

for these applications since there is a large amount of surrounding
documentation but no one example to do exactly what you
need. `pseudocode` enforces the practice of creating an __interface__
specification which:

>  - declares the function signature `def get_issues(repository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]`

Reviewer: ```suggestion
 - Declares the function signature `def get_issues(repository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]`,
```

>  - uses the function's `docstring` to specify what the function should be doing

Reviewer: I don't think 'docstring' should be code formatted, it's not an identifier.

```suggestion
 - Uses the function's docstring to specify what the function should be doing, and
```

>  - automates tests run by `pseudocode` to provide feedback to the LLM of the quality and correctness of the code generated

Reviewer: ```suggestion
 - Automates tests run by `pseudocode` to provide feedback to the LLM of the quality and correctness of the code it generated.
```

│                                                                                                  │
│ def run(repository: str) -> list:                                                                │
│     headers = {'Authorization': f"Bearer {os.environ['GITHUB_TOKEN']}"}                          │
│     url = f"https://api.github.com/repos/{repository}/issues"                                    │
│     since = datetime.datetime.now() - datetime.timedelta(days=10)                                │
│     params = {'since': since.isoformat()}                                                        │
│     response = requests.get(url, headers=headers, params=params)                                 │
│     issues = response.json()                                                                     │
│     return [(i['title'], i['number'], i['created_at']) for i in issues if i['number'] % 2 != 0]  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
What code feedback would you like to provide? (leave empty for approval):
```

> Already after some quick initial feedback we can see that the LLM

Reviewer: ```suggestion
After some quick initial feedback we can already see that the LLM
```

> takes into account the users needs. After user provided feedback the

Reviewer: ```suggestion
takes into account the user's needs. After user provided feedback is accounted for the
```

> automated tests are run. If those tests fail or run into exceptions

Reviewer: ```suggestion
automated tests are run. If those tests fail or run into exceptions,
```

]
```

The final generated code after passing all tests.

```python
import os
import datetime
import requests

def run(repository: str) -> list:
    headers = {'Authorization': f"Bearer {os.environ['GITHUB_TOKEN']}"}
    url = f"https://api.github.com/repos/{repository}/issues"
    since = datetime.datetime.now() - datetime.timedelta(days=10)
    params = {'since': since.isoformat()}
    response = requests.get(url, headers=headers, params=params)
    issues = response.json()
    return [(i['title'], i['number'], i['created_at']) for i in issues if i['number'] % 2 != 0]
```

## Conclusion

There are several key ideas here we want to highlight about why we
think the approaches taken in `pseudocode` are unique.

 * __Autogenerated code needs guidance__. Feedback is generated from the user and the rest is guided by automated testing and exception reporting back to the LLM.

 * Interface design is naturally a high level task which is the key to composable code. Declaring interfaces for LLMs to operate within is important. 

 * Separation of code declarations and generated code. Similar to `.py` and `.pyc` files we should separate the interfaces from the autogenerated code.

> We mentioned earlier that this was an experiment of the interaction

Reviewer: ```suggestion
We mentioned earlier that this was an experiment on the interaction
```

between automated testing, user feedback, and LLMs. We want to explore
this more and already have additional ideas like applying formatting and linting to generated
LLM code via [black](https://black.readthedocs.io/en/stable/),
[ruff](https://github.com/charliermarsh/ruff), and

> [isort](https://pycqa.github.io/isort/). Caching of LLM code generated

Reviewer: ```suggestion
[isort](https://pycqa.github.io/isort/), and caching of generated LLM code
```

> to avoid repetitive API calls and a database of reusable code. Similar

Reviewer: ```suggestion
to avoid repetitive API calls by providing a database of reusable code. Similar
```

> to `pseudocode.pseudo_function(...)` it would be nice to have an

Reviewer: ```suggestion
to `pseudocode.pseudo_function(...)`, it would be nice to have an
```

equivalent for arbitrary files. This feels somewhat like
[cookiecutter](https://cookiecutter.readthedocs.io/en/stable/) on

> steroids

Reviewer: ```suggestion
steroids.
```

---
title: 'Exploring the Impact of Feedback Loops on LLM Code Generation'
published: March 27, 2023
author: chris-ostrouchov

> description: 'As an experiment we wanted to add a feedback loop into LLM generated code via type annotations, docstrings, automated tests, and user feedback. We developed a simple package "pseudoscode" which uses the OpenAI API for generating code.'

Reviewer 4: @bskinn @Katebrack  Should the text use "we" or "I"?  @costrouc is the only author, so I think it should be "I".

Reviewer 5: @ericdatakelly I see your point but I do like 'we' to imply a team. It could be he's the author but several worked on the experiment. IMO, "I" puts a lot of emphasis on the one person. Thoughts @bskinn ?

Reviewer: Since this is a company blog, not a personal one, my inclination is to prefer 'we' in just about every circumstance.

I'm finding it hard to think of a situation where a post would reach the point of publication on the blog, in which it was important to distinguish the voice of an author from that of the company/team.

recent advances have revolutionized the field of natural language
processing (NLP) and artificial intelligence (AI), opening up a world
of possibilities across various domains. [OpenAI's
ChatGPT](https://chat.openai.com/) has taken the world by storm and
shown remarkable ability to generate human-like responses to a wide
range of queries.

These models require training on vast amounts of data and in some
sense can be thought of as a way to summarize and expose the
information in the training dataset. It is important to remember,
however, that the information returned in response to a query is a
statistical answer that does not result from the same logical
reasoning steps humans are capable of. This leads to the so called
hallucination effect that is now being talked about, i.e., when a
confidently written but factually wrong response is given to a set of
queries.

Between discussion sites like StackOverflow and code repositories like
GitHub and GitLab, there is a tremendous amount of source code
available online for potential training. Efforts like [Github
Copilot](https://github.com/features/copilot), show the power of AI

> assisted coding to boost productivity. Our testing has shown however

Reviewer 4: For consistency.
```suggestion
assisted coding to boost productivity. Our testing has shown, however,
```

that while Copilot can greatly assist a developer, it is also prone to
hallucinations inventing new methods that don't exist or sometimes
using outdated APIs that no longer exist in recent versions of the
code.

We decided to experiment to see if we could add a feedback loop into
the process of code generation to reduce model hallucinations and
iterate towards a working solution faster. We also experimented to

> experiment with adding conversational feedback to the results rather

Reviewer 4: ```suggestion
 with adding conversational feedback to the results rather
```

than the auto-complete mechanism that Copilot currently implements
(Note: Copilot just announced a preview release of a [conversational
chat based
mode](https://github.blog/2023-03-22-github-copilot-x-the-ai-powered-developer-experience/).

As an experiment we developed a Python package
[pseudocode](https://github.com/Quansight/pseudocode) which allows
developers to use an LLM chat session to produce correct and tested
code simply by providing code annotations and tests. We believe that
[pseudocode](https://github.com/Quansight/pseudocode) is a "higher
level language" for writing Python code. Below you will find an

> example. We emphasize that the interface is via type annotations and function docstrings and provide easy ways to include automated tests.

Reviewer 4: ```suggestion
example. We emphasize that the interface is via type annotations and function docstrings and provides easy ways to include automated tests.
```

1. We define the code to be generated by providing a function
   signature with type annotations
2. We define tests that must pass
3. We submit these to OpenAI ChatGPT 3.5 turbo API with some
   instructions
4. Any feedback is resubmitted to the ChatGPT session to continue
   refining the code

> 5. If the users provide feedback, then the feedback is sent back to

Reviewer 4: ```suggestion
5. If the user provides feedback, then the feedback is sent back to
```

   ChatGPT and we repeat the process
6. We run the resulting code, and if the tests fail, we resubmit the
   errors to the ChatGPT session

Below we walk through a demonstration of this process.

## Example

Let's use a concrete example of a slightly non-trivial task of getting

> Github issues.

Reviewer 4: ```suggestion
GitHub issues.
```

import typing
import datetime

from pseudocode import pseudo_function

@pseudo_function(review=True)
def get_issues(repository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]:
    """A function to fetch all issues created by

    A function that does the following:
     - assume you have a github token environment variable GITHUB_TOKEN
     - assume that repository is of the form organization/repo
     - use the requests library
     - fetch all github issues from repository in last 10 days
     - only show issue numbers which are odd
     - return a tuple with issue titles, number, and date created

    Examples
    --------
    >>> all([_[1] % 2 == 1 for _ in get_issues("conda/conda")])
    True
    """
    pass

print('github issues', get_issues("conda/conda"))
```

We feel this is a good example because it is hard to remember how to
use the GitHub API exactly and inevitably requires some back and forth
using the [PyGithub](https://github.com/PyGithub/PyGithub) Python
library, API requests, and viewing API documentation. LLMs show promise

for these applications since there is a large amount of surrounding
documentation but no one example to do exactly what you

> need. `pseudocode` enforces the practice of creating an __interface__

Reviewer 4: Is "interface" intended to be bold?

specification which:
 - declares the function signature `def get_issues(repository: str) -> typing.List[typing.Tuple[str, int, datetime.datetime]]`
 - uses the function's `docstring` to specify what the function should be doing
 - automates tests run by `pseudocode` to provide feedback to the LLM of the quality and correctness of the code generated

> When you run the example above, you will see the following.

Reviewer 4: ```suggestion
The specification is essential for helping the LLM avoid hallucinating, and we have found that it is highly effective for the examples we've experimented with. When you run the example above, you will see the following.
```

Reviewer 4: I think this is an important point to be made at this location in the text.

│     url = f"https://api.github.com/repos/{repository}/issues"                                    │
│     since = datetime.datetime.now() - datetime.timedelta(days=10)                                │
│     params = {'since': since.isoformat()}                                                        │
│     response = requests.get(url, headers=headers, params=params)                                 │
│     issues = response.json()                                                                     │
│     return [(i['title'], i['number'], i['created_at']) for i in issues if i['number'] % 2 != 0]  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
What code feedback would you like to provide? (leave empty for approval):
```

Already after some quick initial feedback we can see that the LLM
takes into account the users needs. After user provided feedback the
automated tests are run. If those tests fail or run into exceptions
they will be automatically passed back to the LLM for corrections.

> ```shell

Reviewer 4: Does this show the tests passing?  It's not obvious to me that the code shows pass or fail.

    return [(i['title'], i['number'], i['created_at']) for i in issues if i['number'] % 2 != 0]
```

## Conclusion

There are several key ideas here we want to highlight about why we
think the approaches taken in `pseudocode` are unique.

 * __Autogenerated code needs guidance__. Feedback is generated from the user and the rest is guided by automated testing and exception reporting back to the LLM.

 * Interface design is naturally a high level task which is the key to composable code. Declaring interfaces for LLMs to operate within is important. 

 * Separation of code declarations and generated code. Similar to `.py` and `.pyc` files we should separate the interfaces from the autogenerated code.

We mentioned earlier that this was an experiment of the interaction
between automated testing, user feedback, and LLMs. We want to explore
this more and already have additional ideas like applying formatting
and linting to generated LLM code via
[black](https://black.readthedocs.io/en/stable/),
[ruff](https://github.com/charliermarsh/ruff), and
[isort](https://pycqa.github.io/isort/), and caching of generated LLM
code to avoid repetitive API calls and store a database of reusable

> code. Similar to `pseudocode.pseudo_function(...)`, it would be nice

Reviewer 4: ```suggestion
code. Similar to `pseudocode.pseudo_function(...)` shown below, it would be nice
```

to have an equivalent for arbitrary files. This would provide a feel
somewhat like
[cookiecutter](https://cookiecutter.readthedocs.io/en/stable/) on
steroids.

```python
def test_dockercompose(contents):
    pass

pseudocode.pseudo_file(
  'docker-compose.yaml, 
  'A docker-compose file to generate a running redis service on port 5000 with no password', 
  [test_dockercompose]
)
```

> 

Reviewer 4: ```suggestion
To see the full example, check out this video.
```

---

> title: 'Collaborative code generation with LLMs'

Reviewer 2: maybe - 'Build your own code copilot with LLMs'

Author: Resolved with latest commit.

published: March 27, 2023
author: chris-ostrouchov
description: 'As an experiment we developed a python package pseudocode which much like its name implies allows developers to annotate their code with pseudocode and produce testable code generated by LLMs without the user writing any code.'
category: ["ML", "LLM", "python", "openai"]
featuredImage:
  src: '/posts/openai-pseudocode/pseudocode-screenshot.png'
  alt: 'A screenshot of the pseudocode program running'
hero:

>   imageSrc: /posts/hero-paris.webp

Reviewer 2: this should be changed. I don't think a paris viz makes sense as a hero image for this.

Reviewer: Be aware that the way the header is currently designed, it's necessary to have a dark hero.

Otherwise, the header navigation will be a light gray against a light background, making it very hard to see.

language processing (NLP) and artificial intelligence (AI), opening up
a world of possibilities across various domains. These models have
been trained on vast amounts of data and are capable of generating
human-like responses to a wide range of queries. The developer
community has seen especially good results due to the large amount of
source code available online for training.

Github's Co-pilot is a tool which will autocomplete code for a
developer but lacks the ability to interact with the completation,
provide feedback, and enforce a test suite on the generated code.

As an experiment we developed a python package
[pseudocode](https://github.com/Quansight/pseudocode) which much like
its name implies allows developers to annotate their code with
pseudocode and produce testable code generated by LLMs without the
user writing any code. We hope that
[pseudocode](https://github.com/Quansight/pseudocode) may be a "higher
level language" for writing python code. Below you will find an
example. We emphasize interfaces via type annotations and function
docstrings and provide easy ways to have automated tests.

```python

> from pseudocode import pseudo_function

Reviewer 2: not a great example. 

it would be better to do what you did your live demo. walk through the process of downloading the issues from the gh api and doing a q&a with it and getting the final result. the sum_func is very underwhelming.

@pseudo_function(review=True)
def my_sum_func(a: int, b: int, c: int) -> str:
    """A fun sum function

    The function does the following:
     - adds all arguments together
     - then multiplies by two

    Examples
    --------
    >>> my_sum_func(1, 2, 3)
    12
    """
    pass

my_sum_func(1, 3, 4)
```

> One of the key features of this library is its powerful feedback loop,

Reviewer 2: why did you write this and why do you feel it would be great...
